pulsar-client/src/main/java/org/apache/pulsar/client/impl/schema/generic/GenericSchemaImpl.java

```diff
@@ -41,10 +43,126 @@ protected GenericSchemaImpl(SchemaInfo schemaInfo) {
 
         this.fields = schema.getFields()
                 .stream()
-                .map(f -> new Field(f.name(), f.pos()))
+                .map(f -> new Field(f.name(), f.pos(), convertFieldSchema(f, schemaInfo.getType())))
                 .collect(Collectors.toList());
     }
 
+
+    public static org.apache.pulsar.client.api.Schema<?> convertFieldSchema(Schema.Field f, SchemaType mainType) {
+        return convertFieldSchema(f.schema(), mainType);
+    }
+
+    private static org.apache.pulsar.client.api.Schema<?> convertFieldSchema(Schema schema, SchemaType mainType) {
+        switch (schema.getType()) {
+            case RECORD:
+                return buildStructSchema(schema, mainType);
+            case BYTES:
+                return GenericSchema.BYTES;
+            case LONG:
+                return GenericSchema.INT64;
+            case FLOAT:
+                return GenericSchema.FLOAT;
+            case DOUBLE:
+                return GenericSchema.DOUBLE;
+            case BOOLEAN:
+                return GenericSchema.BOOL;
+            case STRING:
+                return GenericSchema.STRING;
+            case INT:
+                return GenericSchema.INT32;
+            case UNION:
+                // this is very common while representing NULLABLE types
+                // the first entry is "null", the second is the effective type
+                List<Schema> types = schema.getTypes();
+                if (types.size() == 2) {
+                    if (types.stream().anyMatch(s -> s.getType() == Schema.Type.NULL)) {
```

Comment:
`if (types.size() == 2 && types.stream().filter(s -> s.getType() != Schema.Type.NULL)).count() == 1)` ?
removes need for `.orElse` and null check later